packages/ra-ui-materialui/src/input/NumberInput.spec.tsx

```diff
@@ -290,6 +292,47 @@ describe('<NumberInput />', () => {
             });
             expect(onSubmit.mock.calls[0][0].views).toBeNull();
         });
+
+        it('should use custom format function prop', async () => {
+            const ValueInput = props => {
+                const unit = useWatch({ name: 'unit' });
+                const formatFunction = v => (unit === 'radian' ? 1 : 2);
+
+                return (
+                    <NumberInput
+                        format={formatFunction}
+                        {...props}
+                        defaultValue={1}
+                    />
+                );
+            };
+
+            const units = ['degree', 'radian'];
+            const toChoices = items =>
+                items.map(item => ({ id: item, name: item }));
+
+            render(
+                <AdminContext>
+                    <SimpleForm>
+                        <SelectInput source="unit" choices={toChoices(units)} />
+                        <ValueInput {...defaultProps} />
+                    </SimpleForm>
+                </AdminContext>
+            );
+
+            userEvent.click(
+                screen.getByLabelText('resources.undefined.fields.unit')
+            );
+            userEvent.click(screen.getByRole('option', { name: /^radian/i }));
+
+            await waitFor(() => {
```

Comment:
This test is incomplete ; it doesn't show that the value has changed, because it never asserted the value before the change. 

Besides, I don't understand why you have to use a dependent input to test the format prop. Can't you write a simpler test?